<?xml version="1.0"?>
<robot name="xarm5" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:include filename="$(find pickplace)/urdf/xarm5/xarm5_gripper.urdf.xacro"/>

  <!-- Material definitions -->
  <material name="White">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>
  <material name="Silver">
    <color rgba="0.8 0.8 0.8 1.0"/>
  </material>

  <link name="world"/>
  <joint name="world_joint" type="fixed">
    <parent link="world"/>
    <child link="link_base"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  
  <!-- Links -->
  <link name="link_base">
    <inertial>
      <origin xyz="-0.00218 -0.00023 0.08604" rpy="0 0 0" />
      <mass value="1.95" />
      <inertia ixx="0.004805" ixy="-8.33E-06" ixz="0.00026847" iyy="0.0050349" iyz="7.23E-06" izz="0.0025418" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link_base.stl"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link_base.stl"/>
      </geometry>
    </collision>
  </link>

  <link name="link1">
    <inertial>
      <origin xyz="0.00017 0.01881 -0.05409" rpy="0 0 0"/>
      <mass value="2.04"/>
      <inertia ixx="0.0038793" ixy="-0.0001124" ixz="0.0000411" iyy="0.0038791" iyz="-0.0002699" izz="0.0014988"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link1.stl"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link1.stl"/>
      </geometry>
    </collision>
  </link>

  <link name="link2">
    <inertial>
      <origin xyz="0.00035 0.02638 -0.02586" rpy="0 0 0"/>
      <mass value="2.04"/>
      <inertia ixx="0.0038793" ixy="-0.0001124" ixz="0.0000411" iyy="0.0038791" iyz="-0.0002699" izz="0.0014988"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link2.stl"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link2.stl"/>
      </geometry>
    </collision>
  </link>

  <link name="link3">
    <inertial>
      <origin xyz="0.00035 0.02638 -0.02586" rpy="0 0 0"/>
      <mass value="2.04"/>
      <inertia ixx="0.0038793" ixy="-0.0001124" ixz="0.0000411" iyy="0.0038791" iyz="-0.0002699" izz="0.0014988"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link3.stl"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link3.stl"/>
      </geometry>
    </collision>
  </link>

  <link name="link4">
    <inertial>
      <origin xyz="0.00035 0.02638 -0.02586" rpy="0 0 0"/>
      <mass value="2.04"/>
      <inertia ixx="0.0038793" ixy="-0.0001124" ixz="0.0000411" iyy="0.0038791" iyz="-0.0002699" izz="0.0014988"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link4.stl"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link4.stl"/>
      </geometry>
    </collision>
  </link>

  <link name="link5">
    <inertial>
      <origin xyz="0.00035 0.02638 -0.02586" rpy="0 0 0"/>
      <mass value="2.04"/>
      <inertia ixx="0.0038793" ixy="-0.0001124" ixz="0.0000411" iyy="0.0038791" iyz="-0.0002699" izz="0.0014988"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/xarm5/visual/link5.stl"/>
      </geometry>
      <material name="Silver"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pickplace/meshes/end_tool/collision/end_tool.stl"/>
      </geometry>
    </collision>
  </link>

  <!-- Joints -->
  <joint name="joint1" type="revolute">
    <origin xyz="0 0 0.267" rpy="0 0 0" />
    <parent link="link_base" />
    <child link="link1" />
    <axis xyz="0 0 1" />
    <limit lower="-6.283185307179586" upper="6.283185307179586" effort="50" velocity="3.14" />
    <dynamics damping="10" friction="1" />
  </joint>

  <joint name="joint2" type="revolute">
    <origin xyz="0 0 0" rpy="-1.5708 0 0" />
    <parent link="link1" />
    <child link="link2" />
    <axis xyz="0 0 1" />
    <limit lower="-2.059" upper="2.0944" effort="50" velocity="3.14" />
    <dynamics damping="10" friction="1" />
  </joint>

  <joint name="joint3" type="revolute">
    <origin xyz="0.0535 -0.2845 0" rpy="0 0 0" />
    <parent link="link2" />
    <child link="link3" />
    <axis xyz="0 0 1" />
    <limit lower="-3.927" upper="0.19198" effort="30" velocity="3.14" />
    <dynamics damping="5" friction="1" />
  </joint>

  <joint name="joint4" type="revolute">
    <origin xyz="0.0775 0.3425 0" rpy="0 0 0" />
    <parent link="link3" />
    <child link="link4" />
    <axis xyz="0 0 1" />
    <limit lower="-1.69297" upper="3.141592653589793" effort="20" velocity="3.14" />
    <dynamics damping="3" friction="1" />
  </joint>

  <joint name="joint5" type="revolute">
    <origin xyz="0.076 0.097 0" rpy="-1.5708 0 0" />
    <parent link="link4" />
    <child link="link5" />
    <axis xyz="0 0 1" />
    <limit lower="-6.283185307179586" upper="6.283185307179586" effort="20" velocity="3.14" />
    <dynamics damping="3" friction="1" />
  </joint>

  <!-- End Effector -->
  <link name="link_eef" />
  <joint name="joint_eef" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0" />
    <parent link="link5" />
    <child link="link_eef" />
  </joint>
  <xacro:xarm5_gripper prefix="gripper_" parent="link_eef" />

  <!--Camera Definitions-->
  <joint name="d435_camera_joint" type="fixed">
    <parent link="link5"/>
    <child link="d435_camera_link"/>
    <origin xyz="0.07 0 -0.055" rpy="0 0 1.57909"/>
  </joint>
  <link name="d435_camera_link">
    <visual>
      <geometry><box size="0.1 0.02 0.02"/></geometry>
      <material name="red"><color rgba="1 0 0 1"/></material>
    </visual>
    <collision>
      <geometry><box size="0.1 0.02 0.02"/></geometry>
    </collision>
    <inertial>
      <mass value="0.01"/>
      <inertia ixx="0.00001" ixy="0" ixz="0" iyy="0.00001" iyz="0" izz="0.00001"/>
    </inertial>
  </link>
  <gazebo reference="d435_camera_link">
    <sensor name="rgb_camera" type="camera">
      <pose>0 0 0 1.57909 -1.57909 0</pose>
      <always_on>1</always_on>
      <update_rate>30</update_rate>
      <visualize>true</visualize>
      <frame_name>d435_camera_link</frame_name>
      <camera>
        <horizontal_fov>2.0944</horizontal_fov>
        <image>
          <width>1280</width>
          <height>720</height>
          <format>R8G8B8</format>
        </image>
        <lens>
          <intrinsics>
            <fx>369.4</fx>
            <fy>369.4</fy>
            <cx>640.0</cx>
            <cy>360.0</cy>
          </intrinsics>
        </lens>
        <clip>
          <near>0.1</near>
          <far>10.0</far>
        </clip>
        <topic>rgb_camera</topic>
      </camera>
    </sensor>
    <sensor name="rgbd_camera" type="rgbd_camera">
      <pose>0 0 0 1.57909 -1.57909 0</pose>
      <always_on>1</always_on>
      <update_rate>30</update_rate>
      <visualize>true</visualize>
      <frame_name>d435_camera_link</frame_name>
      <camera>
        <horizontal_fov>2.0944</horizontal_fov>
        <image>
          <width>1280</width>
          <height>720</height>
          <format>R8G8B8</format>
        </image>
        <lens>
          <intrinsics>
            <fx>369.4</fx>
            <fy>369.4</fy>
            <cx>640.0</cx>
            <cy>360.0</cy>
          </intrinsics>
        </lens>
        <clip>
          <near>0.1</near>
          <far>10.0</far>
        </clip>
      </camera>
      <depth_camera>
        <output>depth</output>
        <topic>depth_camera</topic>
      </depth_camera>
    </sensor>
  </gazebo>

  <!-- Transmission for ROS Control -->
  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint1">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint2">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint3">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor3">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint4">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor4">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran5">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint5">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor5">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <!--XArm Control Block-->
  <ros2_control name="xarm5" type="system">
    <hardware>
      <plugin>gz_ros2_control/GazeboSimSystem</plugin>
    </hardware>
    <joint name="joint1">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
    <joint name="joint2">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
    <joint name="joint3">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
    <joint name="joint4">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
    <joint name="joint5">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>
  <ros2_control name="gripper" type="system">
    <hardware>
      <plugin>gz_ros2_control/GazeboSimSystem</plugin>
    </hardware>
    <joint name="gripper_drive_joint">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>

  <gazebo>
    <plugin filename="gz_ros2_control-system" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
      <parameters>/home/atharva/assessment_ws/src/pickplace/config/xarm_controller.yaml</parameters>
      <controller_manager_name>controller_manager</controller_manager_name>
    </plugin>
  </gazebo>
</robot>